### DISCUSSION TITLE
Extension Methods

### DISCUSSION THREAD

**Discussion by Guilherme-Cassel** [QuestPDF/QuestPDF#765] [General]
Hello, I'm kinda new to C# (9 months study).

On QuestPDF, is there another way to separate PDF parts in methods (I mean if you do the hole Document.Create in a single method, it would be a nightmare to edit/maintain 😂), if not by static extension methods?

I've made some Reports with QuestPDF already, but all of them are separated with extension methods, which is okay, but I really wanted to make the Report Class non-static and work with Class Fields, so it gets more readable and easier to maintain on the future.

On C# you can't place extension methods on a non-static class, and you cant create fields on a static one, so, is there another clean way to make almost the same, separate The reports parts, like on the image, but without extension methods, like, work with the IContainer where I placed it?

I thought something with Action type but its not very clean to me.

![Captura de tela 2024-01-25 150021](https://github.com/QuestPDF/QuestPDF/assets/150675837/de9537f9-12c4-4889-890e-237ad8099e96)

This ↑ is an example report I made these days, with static class but you can see the problem, I repeatedly pass the same arguments to methods with is not clean at all, I really wanted to have internal fields and then instantiate the class but its impossible because I'm kinda dependant of those IContainer extension methods.

**Comment by MarcinZiabek** [comment]
This approach may become helpful, where you separate document content into separate class methods and reusable layouts into components: https://github.com/QuestPDF/QuestPDF-ExampleInvoice/blob/main/InvoiceDocument.cs

**Reply by Guilherme-Cassel** [reply]
hmm yeah now I got, using the .Element instead of creating a new extension method. 

Thanks, it solved my problem!

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
